src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb

```diff
@@ -473,5 +474,328 @@ End Module
 </compilation>).VerifyDiagnostics(
             Diagnostic(ERRID.ERR_CaseElseNoSelect, "Case Else"))
         End Sub
+
+        <Fact>
+        Public Sub LockType_InSyncLock()
+            Dim source = "
+Module Program
+    Sub Main()
+        Dim l = New System.Threading.Lock()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Lock
+    End Class
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics(
+"BC42508: A value of type 'System.Threading.Lock' in SyncLock will use likely unintended monitor-based locking. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.
+        SyncLock l
+                 ~
+")
+        End Sub
+
+        <Fact>
+        Public Sub LockType_Generic()
+            Dim source = "
+Module Program
+    Sub Main()
+        Dim l = New System.Threading.Lock(Of String)()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Lock(Of T)
+    End Class
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics()
+        End Sub
+
+        <Fact>
+        Public Sub LockType_Nested()
+            Dim source = "
+Module Program
+    Sub Main()
+        Dim l = New System.Threading.Container.Lock()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Container
+        Public Class Lock
+        End Class
+    End Class
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics()
+        End Sub
+
+        <Fact>
+        Public Sub LockType_WrongNamespace()
+            Dim source = "
+Module Program
+    Sub Main()
+        Dim l = New Threading.Lock()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace Threading
+    Public Class Lock
```

Comment:
>Lock [](http://example.com/codeflow?start=17&length=4)

For completeness, consider adding a test for correct namespace, but "wrong" type name.